### PR TITLE
supported to kafka 0.9.0.0 and above versions

### PR DESCRIPTION
**Pull request by hariss63**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -107,8 +113,8 @@
                 <artifactId>maven-compiler-plugin</artifactId>
                 <version>2.0</version>
                 <configuration>
-                    <source>1.5</source>
-                    <target>1.5</target>
+                    <source>1.7</source>
+                    <target>1.7</target>
```
Comment (malakasilva) [inline pom.xml]: Why not 1.8?

File: src/main/java/org/wso2/carbon/connector/KafkaConnectConstants.java [hunk 2]
```diff
@@ -21,71 +21,163 @@
...
-    public static final String KAFKA_BUFFER_SIZE = "kafka.sendBufferSize";
-    public static final String KAFKA_REQUEST_TIMEOUT = "kafka.requestTimeout";
-    public static final String KAFKA_ENQUEUE_TIMEOUT = "kafka.enqueueTimeout";
+    public static final String KAFKA_VALUE_SERIALIZER_CLASS = "kafka.valueSerializerClass";
+    public static final String KAFKA_ACKS = "kafka.acks";
+    public static final String KAFKA_BUFFER_MEMORY = "kafka.bufferMemory";
+    public static final String KAFKA_COMPRESSION_TYPE = "kafka.compressionType";
+    public static final String KAFKA_RETRIES = "kafka.retries";
+    public static final String KAFKA_SSL_KEY_PASSWORD = "kafka.sslKeyPassword";
+    public static final String KAFKA_SSL_KEYSTORE_LOCATION = "kafka.sslKeystoreLocation";
```
Comment (malakasilva) [inline src/main/java/org/wso2/carbon/connector/KafkaConnectConstants.java]: Why we need to prefix kafka?

Comment (hariss63) [inline src/main/java/org/wso2/carbon/connector/KafkaConnectConstants.java]: we gave the property name as kafka.<something> in the template.

File: pom.xml (hunk 1 above)
Comment (malakasilva) [inline pom.xml]: Did we change this?

File: src/main/java/org/wso2/carbon/connector/KafkaProduceConnector.java [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+     * @param partitionNo    The partition Number of the broker.
+     * @param key            The key.
+     * @param message        The message.
+     * @throws ConnectException The Exception while create the connection from the Connection pool.
+     */
+    private void sendWithPool(MessageContext messageContext, String topic, String partitionNo, String key,
+            String message) throws ConnectException {
+
+        KafkaProducer<String, String> producer = KafkaConnectionPool.getConnectionFromPool();
+        if (producer == null) {
```
Comment (malakasilva) [inline src/main/java/org/wso2/carbon/connector/KafkaProduceConnector.java]: Here we create all the connections and send the message. Due to that we may loose the first massage. 
So either we have to do this asynchronously or just create one connection and only add to the pool at the end.

File: src/main/java/org/wso2/carbon/connector/KafkaConnectionPool.java [hunk 4]
```diff
@@ -90,21 +67,21 @@ private synchronized boolean checkIfConnectionPoolIsFull(MessageContext messageC
...
     /**
      * Get the connection from connection pool.
      *
-     * @return the connection
+     * @return the connection.
      */
-    public synchronized Producer<String, String> getConnectionFromPool() {
-        Producer<String, String> connection = null;
+    public static synchronized KafkaProducer<String, String> getConnectionFromPool() {
+        KafkaProducer<String, String> connection = null;
```
Comment (malakasilva) [inline src/main/java/org/wso2/carbon/connector/KafkaConnectionPool.java]: Why do we need this variable. You can simply return. Also have some debug logs.
Have you done research with ConcurrentLinkedQueue?
In this case you may not need to check the size.
https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ConcurrentLinkedQueue.html#poll()

File: src/main/java/org/wso2/carbon/connector/KafkaProduceConnector.java (hunk 3 above)
Comment (hariss63) [inline src/main/java/org/wso2/carbon/connector/KafkaProduceConnector.java]: Here we check the existence of the connection and if there is no connection we need to create the connection pool and get the kafka connection from the pool. Then only we send the message using that connection.

File: src/main/java/org/wso2/carbon/connector/KafkaConnectionPool.java (hunk 4 above)
Comment (hariss63) [inline src/main/java/org/wso2/carbon/connector/KafkaConnectionPool.java]: yes, I did the research on concurrentLinkedQueue and when I use those I need to implement all the methods for KafkaProducer but when I search on vector It is more thread safe implementation by its own.
https://stackoverflow.com/questions/23246059/java-vector-thread-safety

**Pull request merged**